### PR TITLE
feat: merkle tree verifier implementation to support all numbers of leaves

### PR DESCRIPTION
**Pull request by SweeXordious**
## Overview

Closes https://github.com/celestiaorg/blobstream-contracts/issues/249

The implementation is taken from: https://github.com/celestiaorg/celestia-core/blob/0498541b8db00c7fefa918d906877ef2ee0a3710/crypto/merkle/proof.go#L166-L197

## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

## Summary by CodeRabbit

The existing bullet-point list is still valid based on the provided information. No changes are required.

### CODE REVIEW

**Comment by SweeXordious** [comment]
@coderabbitai review

File: src/lib/tree/binary/BinaryMerkleTree.sol [hunk 1]
```diff
@@ -42,65 +42,53 @@ library BinaryMerkleTree {
...
+    /// @param _data The slice that we want to select data from.
+    /// @param _begin The beginning of the range (inclusive).
+    /// @param _end The ending of the range (exclusive).
+    /// @return _ the sliced data.
+    function slice(bytes32[] memory _data, uint256 _begin, uint256 _end) internal pure returns (bytes32[] memory) {
+        bytes32[] memory out = new bytes32[](_end-_begin);
+        for (uint256 i = _begin; i < _end; i++) {
+            out[i] = _data[i];
+        }
+        return out;
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply src/lib/tree/binary/BinaryMerkleTree.sol]: [946d5ee](https://github.com/celestiaorg/blobstream-contracts/pull/253/commits/946d5ee0c091c4471fedb0fa6d3430b5abea5ba4)

File: src/lib/tree/binary/test/BinaryMerkleTree.t.sol [hunk 2]
```diff
@@ -101,6 +101,36 @@ contract BinaryMerkleProofTest is DSTest {
...
+        sideNodes[0] = 0x4f35212d12f9ad2036492c95f1fe79baf4ec7bd9bef3dffa7579f2293ff546a4;
+        sideNodes[1] = 0x6bcf0e2e93e0a18e22789aee965e6553f4fbe93f0acfc4a705d691c8311c4965;
+        sideNodes[2] = 0x4301a067262bbb18b4919742326f6f6d706099f9c0e8b0f2db7b88f204b2cf09;
+
+        uint256 key = 2;
+        uint256 numLeaves = 8;
+        BinaryMerkleProof memory proof = BinaryMerkleProof(sideNodes, key, numLeaves);
+        bytes memory data = hex"03";
+        bool isValid = BinaryMerkleTree.verify(root, proof, data);
+        assertTrue(isValid);
```
Comment (evan-forbes) [inline src/lib/tree/binary/test/BinaryMerkleTree.t.sol:131]: can we also include a test where we expect the verification to fail?

**Review by evan-forbes — COMMENTED** [review]
left one comment for adding tests for false positives, other than that this implementation makes sense to me

File: src/lib/tree/binary/test/BinaryMerkleTree.t.sol (hunk 2 above)
Comment (SweeXordious) [reply src/lib/tree/binary/test/BinaryMerkleTree.t.sol:131]: [a8855d0](https://github.com/celestiaorg/blobstream-contracts/pull/253/commits/a8855d08ae9353acafad23028e983b1b8d581025)

Comment (SweeXordious) [reply src/lib/tree/binary/test/BinaryMerkleTree.t.sol:131]: It was a good idea to add invalid proofs :+1: Found a bug :D

Comment (SweeXordious) [reply src/lib/tree/binary/test/BinaryMerkleTree.t.sol:131]: actually not a bug, but an edge case:

Assume you have a merkle tree of 5 leaves. When creating the proof, if you set the `number of leaves = [5, 8]`, the proof going to pass even if `{6, 7, 8}` are invalid values because with those numbers of leaves, the verifier will behave the same. 

Verified with the implementation of tendermint and it also has the same behaviour.

**Review by evan-forbes — DISMISSED** [review]
lgtm. given the importance of this lib I think it would be worth setting up a fuzzer.

**Comment by SweeXordious** [comment]
> lgtm. given the importance of this lib I think it would be worth setting up a fuzzer.

https://github.com/celestiaorg/blobstream-contracts/issues/256

File: src/lib/tree/binary/test/BinaryMerkleTree.t.sol (hunk 2 above)
Comment (adlerjohn) [reply src/lib/tree/binary/test/BinaryMerkleTree.t.sol:131]: Could you expand on this? What does `number of leaves = [5, 8]` mean? That's a range, not a number.

Comment (SweeXordious) [reply src/lib/tree/binary/test/BinaryMerkleTree.t.sol:131]: My bad, i meant any number from that range will result in a valid proof. The verifier won't complain about it.

So if you have a tree of 5 leaves, and in the proof you set the total number to 7, the proof is still considered valid

**Review by evan-forbes — APPROVED** [review]
refreshing this approve

File: src/lib/tree/binary/test/BinaryMerkleTree.t.sol (hunk 2 above)
Comment (SweeXordious) [reply src/lib/tree/binary/test/BinaryMerkleTree.t.sol:134]: https://github.com/celestiaorg/blobstream-contracts/issues/259

**Comment by SweeXordious** [comment]
We will go on and merge this PR as it's blocking mocha's deployment. The only concern about this implementation is documented in https://github.com/celestiaorg/blobstream-contracts/issues/259. We can open subsequent PRs if any issues arise with it.

File: src/lib/tree/binary/BinaryMerkleTree.sol [hunk 3]
```diff
@@ -42,65 +42,59 @@ library BinaryMerkleTree {
...
-
-        // Determine if the next hash belongs to an orphan that was elevated. This
-        // is the case IFF 'stableEnd' (the last index of the largest full subtree)
-        // is equal to the number of leaves in the Merkle tree.
-        if (stableEnd != proof.numLeaves - 1) {
-            if (proof.sideNodes.length <= height - 1) {
-                return false;
+        if (numLeaves == 1) {
+            if (sideNodes.length != 0) {
+                revert("unexpected inner hashes");
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply src/lib/tree/binary/BinaryMerkleTree.sol:63]: I'll create an issue to return error codes instead, thanks 👍
